force-app/main/default/classes/DMLOperations.cls

```diff
@@ -180,15 +245,24 @@ public with sharing class DMLOperations {
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+        List<Contact> upsertList = new List<Contact>();
 		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
+        for(Contact eachContact :contactsToUpsert){
+            // Extract the Account Name from Contact's LastName
+            String lastName = eachContact.LastName;
+            System.debug(lastName);
+
+            // Get the account based on the name or create one if it doesn't exist
+            Account acc = upsertAccount(eachContact.LastName);
+            System.debug(eachContact.Account);
+            
+            // Link the Contact to its Account using the Account Id
+            eachContact.AccountId = acc.Id;
+            upsertList.add(eachContact);
+            System.debug(upsertList);
+        }
```

Comment:
`upsertAccount` performs SOQL + DML inside of this for loop. You need to use single queries and maps you build outside of the main loop here to identify duplicates and determine if you should create new accounts/contacts or use existing. Then, do one DML statement outside of the loop.

Reply:
Implemented approach posted in Slack.

Reply:
👍